### PR TITLE
remove torch.inference_mode from predict/eval units

### PR DESCRIPTION
**Pull request by edward-io**
Please read through our [contribution guide](https://github.com/pytorch/tnt/blob/main/CONTRIBUTING.md) prior to creating your pull request.

Summary:

`torch.inference_mode()` has some problems with distributed collectives, and users may want gradients during eval.  Using `torch.no_grad()` instead and add docs so users can add decorate their step with `inference_mode` if they prefer.

Downside: the tensors returned from dataloader are not covered under `eval_step`

Test plan:

Fixes #{issue number}

### CODE REVIEW

**Comment by edward-io** [comment]
Opened https://github.com/pytorch/pytorch/issues/86830 to track issues with inference_mode.
